### PR TITLE
DependencyGraphBuilder: Add a sanity check for illegal references

### PR DESCRIPTION
**Pull request by oheger-bosch**
cfa52f5 removed a check for invalid references to dependencies from
ProjectAnalyzerResult, because when working with dependency graphs the
conditions could not always be satisfied. This PR re-introduces an
equivalent check on DependencyGraphBuilder. The check is executed per
default when the build() function is called. It can be disabled
though, which is needed for some corner cases, especially by
DependencyGraphConverter.

### CODE REVIEW

File: model/src/test/kotlin/utils/DependencyGraphBuilderTest.kt [hunk 1]
```diff
@@ -173,9 +175,75 @@ class DependencyGraphBuilderTest : WordSpec({
...
+                .addDependency("s", depLog)
+                .build(checkReferences = false)
+        }
     }
 })
 
+/**
+ * A special namespace used by dependencies, for which [PackageRefDependencyHandler] should not create a package.
+ */
+private const val NS_NO_PACKAGE = "no-package"
```
Comment (sschuberth) [inline model/src/test/kotlin/utils/DependencyGraphBuilderTest.kt:245]: Nit: Maybe spell this out as `NO_PACKAGE_NAMESPACE`?

**Review by MarcelBochtler — CHANGES_REQUESTED** [review]

File: model/src/main/kotlin/utils/DependencyGraphBuilder.kt [hunk 2]
```diff
@@ -103,6 +104,12 @@ class DependencyGraphBuilder<D>(
      */
     private val dependencyIds = mutableListOf<Identifier>()
 
+    /**
+     * A set storing the identifiers of all package dependencies with no issues. This is used to check whether there
+     * references to packages not contained in this builder's list of packages.
```
[the remark this answers is not in the document]
Comment (oheger-bosch) [reply model/src/main/kotlin/utils/DependencyGraphBuilder.kt]: The verb was missing. Fixed.

**Review by MarcelBochtler — APPROVED** [review]
